cpp/src/arrow/dataset/scanner_test.cc

```diff
@@ -922,6 +923,22 @@ std::ostream& operator<<(std::ostream& out, const TestScannerParams& params) {
   return out;
 }
 
+// An InMemoryFragment subclass that tracks the calls to ClearCachedMetadata()
+class ClearCachedMetadataFragment : public InMemoryFragment {
+ public:
+  using InMemoryFragment::InMemoryFragment;
+
+  Status ClearCachedMetadata() override {
+    metadata_clear_count_.fetch_add(1);
```

Comment:
```suggestion
    RETURN_NOT_OK(InMemoryFragment::ClearCachedMetadata());
    metadata_clear_count_.fetch_add(1);
```

Reply:
Maybe we should invoke the actual `ClearCachedMetadata()` to do some real clearing?